### PR TITLE
Documenting some parts of gramlib/grammar.ml

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** documentation

We document the following parts of the gramlib/camlp5 parsing engine:
- the "recovery" mecanism (which sometimes bypasses the intent of the user)
- the main parsing start/continue loop
- the algorithm to insert a new rule

Any further suggestions?

### CODE REVIEW

File: gramlib/grammar.ml [hunk 1]
```diff
@@ -166,7 +168,7 @@ and ('self, 'trec, 'a) ty_symbol =
 | Sopt : ('self, 'trec, 'a) ty_symbol -> ('self, 'trec, 'a option) ty_symbol
 | Sself : ('self, mayrec, 'self) ty_symbol
 | Snext : ('self, mayrec, 'self) ty_symbol
-| Snterm : 'a ty_entry -> ('self, norec, 'a) ty_symbol
+| Snterm : 'a ty_entry -> ('self, norec, 'a) ty_symbol (* norec but the entry can nevertheless introduce a loop with the current entry*)
```
Comment (ejgallego) [inline gramlib/grammar.ml:171]: Nit, usually if you put this in a new line is a bit more readable.

**Review by ejgallego — APPROVED** [review]
Thanks @herbelin , this really helps! I will leave until tomorrow in case other people want to proofread @ppedrot

**Review by ppedrot — APPROVED** [review]
No particular comment.

**Comment by ejgallego** [comment]
Failure is due to a flaky test.

**Comment by ejgallego** [comment]
@coqbot: merge now

**Pull request merged**